packages/MarkdownEditor-Core.package/MarkdownInlineParser.class/instance/interpretTokens.st

```diff
@@ -0,0 +1,30 @@
+parsing
+interpretTokens
+
+	| currentToken |
+	self attributes: OrderedCollection new.
+	currentToken := self tokens first.
+	[true] whileTrue: [(currentToken isDelimiter and: [currentToken isCloser]) 
+		ifTrue: [self findOpenerFor: currentToken 
+				ifFound: [:opener | | attribute length |
+					attribute := currentToken attributeStartingAt: opener.
+					self attributes add: attribute.
+					attribute isBold 
+						ifTrue: [length := 2]
+						ifFalse: [length := 1].
+					opener length > length 
+						ifTrue: [self tokens 
+							add: (opener cutFromEnd: length) asStringToken 
+							after: opener]
+						ifFalse: [self tokens at: (self tokens indexOf: opener) put: opener asStringToken].
+					currentToken length > length 
+						ifTrue: [| cut | 
+							cut := (currentToken cutFromStart: length) asStringToken.
+							self tokens 
+								add: cut 
+								before: currentToken.
+							currentToken := cut]
+						ifFalse: [currentToken := self tokens at: (self tokens indexOf: currentToken) put: currentToken asStringToken]]].
+	currentToken := self tokens 
+		after: currentToken
+		ifAbsent: [^ self]]
```

Comment:
This is the method which I think needs to be refactored somehow. Currently I'm not sure yet how. If there are any questions about it please ask. I know it's not quite easy to understand (which is why it needs refactoring...). 

One thing I can tell you is that we don't use the built in squeak iterating over collections because we are modifying the collection while iterating over it. Because of that we need to treat it like a doubly linked list. We can do that by having a pointer to the `currentToken` we are looking at and then using `after: currentToken` to get the next one and using `add:after:` and `add:before:` to insert into the collection. 

One more thing that makes it even more complicated is that we sometimes have to change the class of the `currentToken`. For this we obviously need a new object which means the `currentToken` needs to be replaced in the collection and also we need to change our pointer to it (line 27) because the old object is no longer in the collection so we couldn't use `after:` to get the next one. 

In addition to that we sometimes don't want to look at the next token but instead look at the same token again (when a token is longer that 2 * we need to cut off the first 2 * and then look at the rest again to see if we can find another opener for it). In this case we simply set the `currentToken` pointer to the StringToken we insert before the `currentToken` anyways (line 26) so that `after:` will give is the previous `currentToken` once again.

Reply:
Some other problems I'm seeing here that there are many operations which could be done by the Tokens or Attributes. For instance, you ask a closer "Here is your opener, would you please generate an attribute from it?" And later, "Hi opener, how long are you?" and either "Looks like you are consumed. Please make yourself comfortable in your second life as StringTocken" or "Oh, you have so much potential, but I only need `length` characters from you. Please cut yourself a bit".
(that good dark very quick...)
I have not fully thought about it, but maybe you could do lots of the coordination in the tokens themselves. Like, the opener cuts itself (even to zero) and returns a remainder (either the unmatched thing, or the string token). They already have the knowledge how many characters are needed (btw; you ask the attribute and generate magic numbers -> polymorphism).

Or, you could let the tokens remember how many characters were consumed.

Reply:
I agree this would be better but I don't quite know how exactly to do it.

Reply:
We changed quite a lot. Check it out :)